### PR TITLE
Table Filters for scalar column, ALLOW FILTERING, WARNING

### PR DESCRIPTION
**Pull request by Yuqi-Du**
Table API filters against scalar columns.
`$eq, $ne, $gt, $lt, $gte, $lte, $in`

- Against 15 column types we currently have (blob is a TODO)

`text/int/timestamp/ascii/date/time/boolean/varint/tinyint/decimal/smallint/double/bigint/float/duration`
I have added java doc of how different filters work against different column types, (There are multiple special cases in DB side).
For convenience, reviewer can also refer to this [metrics table](https://docs.google.com/document/d/1Q3uOPuuQBfONrLDd9FiBwXHUV2gMqh4QH8yimoj13Gc/edit#heading=h.2p4zsn8kq88c).

- Added filter analysis process for ALLOW FILTERING and WARNING. Reviewer can refer to the above metrics, or java docs to see how ALLOW FILTERING apply on different use cases.

- Added IT for combination usages of [different filters + different columns types + Index situation] 

TODO
EJSON $binary for blob type. [#1480 ]
Primary key filtering corner case. [#1481 ]

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java
Comment (Yuqi-Du) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:308]: @tatu-at-datastax  @amorton 
This is the quick special handling we talked about last week. 
To unblock table filter against non-numeric column.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/BooleanTableFilter.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/BooleanTableFilter.java:3]: Is this correct? Name implies "Boolean", similarly type parameter -- not number?

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:22]: Is this also expected to be used for "nin"? If so, should have Javadoc.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:29]: Is there expectation that `NIN` will be added? If not, is there any need for `Operator` enumeration, field?

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:71]: typo: "No" -> "Not"

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:22]: Good point, currently it is designed for both $in and $nin, so same class, different operator.
Let me add more docs.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/BooleanTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/BooleanTableFilter.java:3]: yes, thanks!

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:29]: yes, will be added. [Ticket](https://github.com/stargate/data-api/issues/1478)

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:71]: Thanks!

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:26]: once merged, we will work on this refactor https://github.com/stargate/data-api/issues/1485

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java:207]: Use the APiException pls, there is an existing for unknown column. Replace the TABLE_INVALID_FILTER with something more meaninfull,  COMPARISION_ON_DURATION_NOT_SUPPORTED

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:61]: nit: pls remove my name from the comment, I use it for comments I have written

Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:61]: remove my name from the comment pls

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:57]: use ApiException errors , we should copy the ones we have int he DOCUMENT scope for unknown column and unsupported type into the FILTER scope

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:72]: can you clarify what modified means, do we mean the interface will be changed ?

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java:69]: this is a bug - it will set the allow filtering on the CqlOperations object that is used everytime the build() function is called. 

add a copy consturctor to the CqlOptions that takes a CqlOperations as a param and copies the contents, then create a new instance of CqlOperations in this function if we need to allow filtering and use that for the read attempt.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java:83]: not sure on this, certainly the builder is doing the right thing doing the analysis, just nto sure about adding the warnings like this. Nothingt to change for now, just commenting from our discussion

**Review by amorton — CHANGES_REQUESTED** [review]

**Review by amorton — CHANGES_REQUESTED** [review]
There is a bug in TableReadAttemptBuilder, and we some nit picks. 

We will refactor the WhereCqlClauseAnalyser afrer mreged, linked a ticket for this

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:61]: removed

Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:61]: removed

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java:207]: please refer to this comment.
https://github.com/stargate/data-api/pull/1448#discussion_r1783614162

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:72]: added java docs

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java:69]: Solved. Thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:57]: convert UNKNOWN_TABLE_COLUMNS to ERROR V2.
There are other v1 usages, different types, will refactor them in following ticket.

Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:60]: both of those codes are in the error.yaml file here https://github.com/stargate/data-api/blob/main/src/main/resources/errors.yaml#L123

File: src/main/resources/errors.yaml
Comment (amorton) [inline src/main/resources/errors.yaml:136]: is this correct to duplicate the code ?

File: src/main/java/io/stargate/sgv2/jsonapi/exception/FilterException.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/FilterException.java:17]: Both to these are in the yaml file here https://github.com/stargate/data-api/blob/main/src/main/resources/errors.yaml#L123

there were added as place holders during the dev, now dont need them so they can be removed and the entries removed from errors.yaml

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java:207]: -1 on added a new V1 error code, can we remove this pls

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:60]: created ticket so this does not get missed https://github.com/stargate/data-api/issues/1488

feels sloppy to change 1 and not the other below.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java:66]: only create a new options when needed

**Review by amorton — CHANGES_REQUESTED** [review]
only clone the CqlOptions when we need to

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/TableReadAttemptBuilder.java:66]: thanks, solved.

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java:207]: Thanks, removed and change to use v2 ERROR_CODE INVALID_FILTER.

File: src/main/java/io/stargate/sgv2/jsonapi/exception/FilterException.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/FilterException.java:17]: thanks, removed

File: src/main/resources/errors.yaml
Comment (Yuqi-Du) [reply src/main/resources/errors.yaml:136]: solved, thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/InTableFilter.java:60]: removed MULTIPLE_ID_FILTER, FIELDS_LIMIT_VIOLATION.

Added following new v2 codes, to resolve [#1488]
```
  # FILTER request errors: filter against unknown columns
  - scope: FILTER
    code: UNKNOWN_TABLE_COLUMNS
    title: Unknown table columns in filter
    body: |-
      Only columns defined in the table schema can be filtered.
      
      The table ${keyspace}.${table} defines the columns: ${allColumns}.
      The request included the following unknown columns: ${unknownColumns}.
      
      Resend the request using only defined columns.
  # FILTER request errors: filter against a column type that is not supported
  - scope: FILTER
    code: UNSUPPORTED_COLUMN_TYPES
    title: Unsupported table column data types in filter
    body: |-
      Only supported column types can be filtered
      
      The table ${keyspace}.${table} defines the columns: ${allColumns}.
      The request included the following columns that have unsupported data types: ${unsupportedColumns}.
      
      Resend the request using only supported column types.  
```

**Review by amorton — APPROVED** [review]

**Comment by vkarpov15** [comment]
I still get a warning when sending the following command:

```
{"findOne":{"filter":{"vehicleId":"670ae2df8b2f8380b95670f0"},"options":{}}}
```

Response:

```
{
  status: {
    warnings: [
      'The filter includes columns that are not indexed. \n' +
        '\n' +
        'The table demo.reviews has the primary key: "_id"(text).\n' +
        'And has indexes on the columns: [None].\n' +
        'The request filtered on the un-indexed columns: "vehicleId".\n' +
        '\n' +
        'The query was executed without taking advantage of the primary key or indexes on the table, this can have performance implications on large tables.\n' +
        '\n' +
        'See documentation at XXXX for best practices for filtering.'
    ]
  },
  data: {
    document: {
      createdAt: 1728766687851,
      rating: 5,
      _id: '670ae2df8b2f8380b95670f6',
      text: 'This is a review that must have length greater than 30. 5',
      vehicleId: '670ae2df8b2f8380b95670f0',
      userId: '670ae2df8b2f8380b95670ef',
      updatedAt: 1728766687851
    }
  },
  errors: undefined
}
```

I created index as follows:

```
{"addIndex":{"column":"vehicleId","indexName":"vehicleId"}}
```

Is this expected?

**Pull request merged**
